### PR TITLE
[Bignum] Bignumber test mul inv named moduli

### PR DESCRIPTION
**Pull request by xkqian**
## Description

Add one test function to test the mont multiplication and inverse after setting up the named curves moduli.
Get A as an input, calculate A_inv with mbedtls_mpi_mod_inv(). Then call mbedtls_mpi_mod_mul() to calculate A * A_inv and check if the result is 1
Add test cases for the brainpool curves and curves Mbedtls supported. 

## Gatekeeper checklist

- [x] **changelog**  not required
- [x] **backport** not required
- [x] **tests**  not required

### CODE REVIEW

File: tests/suites/test_suite_ecp.function [hunk 1]
```diff
@@ -1433,3 +1433,69 @@ exit:
...
+        ASSERT_ALLOC(rX_raw, limbs);
+        TEST_EQUAL(0, mbedtls_mpi_mod_residue_setup(&rX, &m, rX_raw, limbs));
+
+        /* Get inverse of A mode m, and multiply it with itself,
+         * to see whether the result equal to '1' */
+        TEST_EQUAL(0, mbedtls_mpi_mod_inv(&A_inverse, &A, &m));
+        TEST_EQUAL(mbedtls_mpi_mod_mul(&rX, &A, &A_inverse, &m), 0);
+
+        ASSERT_ALLOC(bufx, limbs);
+        TEST_EQUAL(mbedtls_mpi_mod_write(&rX, &m, (unsigned char *)bufx,
```
Comment (minosgalanakis) [inline tests/suites/test_suite_ecp.function:1482]: I think the code style needs that to be:

TEST_EQUAL(mbedtls_mpi_mod_write(&rX, &m, (unsigned char *)bufx,
```suggestion
        TEST_EQUAL(mbedtls_mpi_mod_write(&rX, &m, (unsigned char *) bufx,
```

File: tests/suites/test_suite_ecp.data [hunk 2]
```diff
@@ -1026,3 +1026,195 @@ ecp_mod_setup:"fffffffffffffffffffffffe26f2fc17f69466a74defd8d":MBEDTLS_ECP_DP_C
 ecp_setup_negative_test #28 Invalid Curve Type
 depends_on:MBEDTLS_ECP_DP_CURVE448_ENABLED
 ecp_mod_setup:"fffffffffffffffffffffffe26f2fc17f69466a74defd8d":MBEDTLS_ECP_DP_NONE:MBEDTLS_ECP_MOD_SCALAR:MBEDTLS_ERR_ECP_BAD_INPUT_DATA
+
+ecp_mul_inv #1 MBEDTLS_ECP_MOD_SCALAR(MBEDTLS_ECP_DP_SECP192R1)
+depends_on:MBEDTLS_ECP_DP_SECP192R1_ENABLED
+ecp_mod_mul_inv:"0000000000000000000000000000152d02c7e14af67fe0bf":MBEDTLS_ECP_DP_SECP192R1:MBEDTLS_ECP_MOD_SCALAR:0
```
Comment (minosgalanakis) [inline tests/suites/test_suite_ecp.data:1032]: How is that data generated? Will there be follow-up commits which will enable it on the python test framework?

Comment (yanesca) [reply tests/suites/test_suite_ecp.data:1032]: I think that these tests are not very convenient to do in the test framework. We would need to extend the framework to accommodate them. I think it is easier from both writing, review and maintenance perspective if we just go for manual test cases.

Comment (yanesca) [reply tests/suites/test_suite_ecp.data:1032]: If we want to capture how these were generated, I suggest we add a comment in the .data file to the test case.

**Comment by xkqian** [comment]
One interesting thing: the Internal CI failed on windwos about the test case ecp_mul_inv **#40, #41,and #42**, which are `MBEDTLS_ECP_DP_SECP224K1`, scalar mode.  But all of them passed in linux. Strange...
I compared the data, take #40 as one examle, "`534a7b77a3ec2d7a8b5771db248ff31da25e928223b638bf9bb8d98b`" totally has 224 bits, and it should be less than `2^224`.   and the N is:
static const mbedtls_mpi_uint secp224k1_n[] = {
    MBEDTLS_BYTES_TO_T_UINT_8(0xF7, 0xB1, 0x9F, 0x76, 0x71, 0xA9, 0xF0, 0xCA),
    MBEDTLS_BYTES_TO_T_UINT_8(0x84, 0x61, 0xEC, 0xD2, 0xE8, 0xDC, 0x01, 0x00),
    MBEDTLS_BYTES_TO_T_UINT_8(0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00),
    MBEDTLS_BYTES_TO_T_UINT_8(0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00),
};
which is `2^224 +...`  and bigger than "`534a7b77a3ec2d7a8b5771db248ff31da25e928223b638bf9bb8d98b`"
Not sure whether the `mbedtls_mpi_core_lt_ct`() failed...

**Comment by gilles-peskine-arm** [comment]
@xkqian Looks like a bug on 32-bit platforms. The 32-bit tests on Linux also failed.

**Comment by gilles-peskine-arm** [comment]
224 is a multiple of 32 but not 64 so it's not a surprise that a 224-bit curve could trigger problems that are specific to 32-bit machines. In addition, secp224k1 is peculiar: unlike all the other curves (including secp224r1), public points are one bit larger: they can get up to slightly _more_ than 2^224 — for other curves both private and public values are less than 2^n. So look for a bug where you try to fit a public point in the same amount of memory as a private point: that works with all curves except secp224k1, that works on 64-bit for secp224k1 because you have room for 256 bits anyway, but it sometimes overflows on 32-bit with secp224k1.

**Comment by xkqian** [comment]
> 224 is a multiple of 32 but not 64 so it's not a surprise that a 224-bit curve could trigger problems that are specific to 32-bit machines. In addition, secp224k1 is peculiar: unlike all the other curves (including secp224r1), public points are one bit larger: they can get up to slightly _more_ than 2^224 — for other curves both private and public values are less than 2^n. So look for a bug where you try to fit a public point in the same amount of memory as a private point: that works with all curves except secp224k1, that works on 64-bit for secp224k1 because you have room for 256 bits anyway, but it sometimes overflows on 32-bit with secp224k1.

Shall I file a bug about this?
On 32 bit platform, just as what you described, it failed in the line` if (**p_limbs != N->limbs** || !mbedtls_mpi_core_lt_ct(p, N->p, N->limbs))` in function mbedtls_mpi_mod_residue_setup(),  which will compare the limbs from input(224 bits, 7 limbs) with the limbs of public point N(225 bits, 8 limbs).  Once mismatch, the function return failure.

**Comment by gilles-peskine-arm** [comment]
> Shall I file a bug about this? Or should we provide the 225 bits input?

I haven't looked at the problematic code so I don't know if the problem is in the test code or in the library code.

> with the most significant bit 1, and following up 112 bits '0', and then the random bits make sure it is less than N?

A 225-bit coordinate doesn't mean that the MSB is 1. It means that the range of the value is from 0 to 2^224+a for some a that's (much) smaller than 2^224. The functions that take `mbedtls_mpi` should accept smaller inputs since leading zeros may be omitted in `mbedtls_mpi`. On the other hand, this may be a problem with the new code that uses fixed-size MPIs: for secp224k1, private and public values need to have different sizes.

File: tests/suites/test_suite_ecp.data [hunk 3]
```diff
@@ -1187,15 +1187,15 @@ ecp_mod_mul_inv:"d2a10413f48d7bcc18a9b7c53c7914c5302c9c9e48b2eb62":MBEDTLS_ECP_D
 
 ecp_mul_inv #40 MBEDTLS_ECP_MOD_SCALAR(MBEDTLS_ECP_DP_SECP224K1)
 depends_on:MBEDTLS_ECP_DP_SECP224K1_ENABLED
-ecp_mod_mul_inv:"534a7b77a3ec2d7a8b5771db248ff31da25e928223b638bf9bb8d98b":MBEDTLS_ECP_DP_SECP224K1:MBEDTLS_ECP_MOD_SCALAR:0
+ecp_mod_mul_inv:"1000000000000000000000000000075ea446a83291f5136799781cfbd":MBEDTLS_ECP_DP_SECP224K1:MBEDTLS_ECP_MOD_SCALAR:0
```
Comment (yanesca) [inline tests/suites/test_suite_ecp.data:1190]: The reason for the limb difference is that `mbedtls_test_read_mpi_core()` always allocates the minimum limb necessary based on the length of the input. If we want to allocate more limbs without changing the value, we just need to prepend a leading zero.

That means, we don't need to set the leading bit to one and we can test with smaller values as well, not just with values between 2^224 and the order of the curve (the order being n=01 00000000 00000000 00000000 0001DCE8 D2EC6184 CAF0A971)
769FB1F7)

**Review by yanesca — CHANGES_REQUESTED** [review]
Left some feedback inline. As usual, I marked the comments I don't consider blocker as "minor". (It is up to you whether you decide to address them or not.)

File: tests/suites/test_suite_ecp.function [hunk 4]
```diff
@@ -1433,3 +1433,69 @@ exit:
...
+{
+    int ret;
+    mbedtls_mpi_mod_modulus m;
+    mbedtls_mpi_mod_modulus_init(&m);
+    mbedtls_mpi_mod_residue A = { NULL, 0 };
+    mbedtls_mpi_mod_residue A_inverse = { NULL, 0 };
+    mbedtls_mpi_mod_residue rX = { NULL, 0 };
+    mbedtls_mpi_uint *rX_raw = NULL;
+    mbedtls_mpi_uint *p_inverse = NULL;
+    mbedtls_mpi_uint *p = NULL;
```
Comment (minosgalanakis) [inline tests/suites/test_suite_ecp.function:1448]: Assuming the input is `void ecp_mod_mul_inv(char *input_A, int id, int ctype, int iret)`
This should be named as `mbedtls_mpi_uint *A = NULL;`

Generally when we are using an `mbedtls_mpi_uint *` pointer which is read through `mbedtls_test_read_mpi_core` the letter is kept.

    char *input_N ->  mbedtls_mpi_uint *N = NULL;

The inverse can be whatever the name of the pointer is plus  _inverse or _inv

Then the pointer for the raw data can be named as such

`mbedtls_mpi_uint *p_inverse = NULL;` --> `mbedtls_mpi_uint *rA_inv_raw= NULL;`

File: tests/suites/test_suite_ecp.function [hunk 5]
```diff
@@ -1433,3 +1433,69 @@ exit:
...
 /* END_CASE */
+
+/* BEGIN_CASE depends_on:MBEDTLS_TEST_HOOKS */
+void ecp_mod_mul_inv(char *input_A, int id, int ctype, int iret)
+{
+    int ret;
+    mbedtls_mpi_mod_modulus m;
+    mbedtls_mpi_mod_modulus_init(&m);
+    mbedtls_mpi_mod_residue A = { NULL, 0 };
+    mbedtls_mpi_mod_residue A_inverse = { NULL, 0 };
```
Comment (minosgalanakis) [inline tests/suites/test_suite_ecp.function:1444]: I would suggest we keep the residue naming notation consistent.

    mbedtls_mpi_mod_residue rA = { NULL, 0 }; // For input
    mbedtls_mpi_mod_residue rX = { NULL, 0 }; // For result
    mbedtls_mpi_mod_residue rA_inv= { NULL, 0 }; // For inverse input

File: tests/suites/test_suite_ecp.function [hunk 6]
```diff
@@ -1433,3 +1433,69 @@ exit:
...
+
+        ASSERT_ALLOC(rX_raw, limbs);
+        TEST_EQUAL(0, mbedtls_mpi_mod_residue_setup(&rX, &m, rX_raw, limbs));
+
+        /* Get inverse of A mode m, and multiply it with itself,
+         * to see whether the result equal to '1' */
+        TEST_EQUAL(0, mbedtls_mpi_mod_inv(&A_inverse, &A, &m));
+        TEST_EQUAL(mbedtls_mpi_mod_mul(&rX, &A, &A_inverse, &m), 0);
+
+        ASSERT_ALLOC(bufx, limbs);
```
Comment (minosgalanakis) [inline tests/suites/test_suite_ecp.function:1481]: I think that aim at this point is to test that the rX residuse is equal to `one` or `Result` .

For that reason the block bellow can and should be replaced the `ASSERT_COMPARE` test helper macro

    ASSERT_COMPARE(rX->p, 1, one, 1);

File: tests/suites/test_suite_ecp.data [hunk 7]
```diff
@@ -1187,15 +1187,15 @@ ecp_mod_mul_inv:"d2a10413f48d7bcc18a9b7c53c7914c5302c9c9e48b2eb62":MBEDTLS_ECP_D
 
 ecp_mul_inv #40 MBEDTLS_ECP_MOD_SCALAR(MBEDTLS_ECP_DP_SECP224K1)
 depends_on:MBEDTLS_ECP_DP_SECP224K1_ENABLED
-ecp_mod_mul_inv:"1000000000000000000000000000075ea446a83291f5136799781cfbd":MBEDTLS_ECP_DP_SECP224K1:MBEDTLS_ECP_MOD_SCALAR:0
```
[the remark this answers is not in the document]
Comment (xkqian) [reply tests/suites/test_suite_ecp.data]: Sure. I will add back the 3 test cases with leading 1.

File: tests/suites/test_suite_ecp.function [hunk 8]
```diff
@@ -1433,3 +1433,69 @@ exit:
...
+        size_t limbs;
+
+        TEST_EQUAL(mbedtls_test_read_mpi_core(&p, &limbs, input_A), 0);
+        TEST_EQUAL(0, mbedtls_mpi_mod_residue_setup(&A, &m, p, limbs));
+
+        /* Test for limb sizes */
+        TEST_EQUAL(m.limbs, limbs);
+
+        /* Test for validity of moduli by the presence of Montgomery consts */
+
```
[the remark this answers is not in the document]
Comment (xkqian) [reply tests/suites/test_suite_ecp.function:1466]: Per the comments from https://github.com/Mbed-TLS/mbedtls/pull/7479#discussion_r1194923840, remove them.

File: tests/suites/test_suite_ecp.data (hunk 7 above)
[the remark this answers is not in the document]
Comment (yanesca) [reply tests/suites/test_suite_ecp.data]: Thanks, I think one should be enough.

File: tests/suites/test_suite_ecp.function (hunk 4 above)
Comment (xkqian) [reply tests/suites/test_suite_ecp.function:1448]: Just renamed the data, and for this variable, not very the smae, ` mbedtls_mpi_uint *p_inverse = NULL;` --> `mbedtls_mpi_uint *A_inv= NULL;`

File: tests/suites/test_suite_ecp.function (hunk 6 above)
Comment (xkqian) [reply tests/suites/test_suite_ecp.function:1481]: Yes, but the representation of rX is in mont types, have to call mbedtls_mpi_mod_write() to transfer it to the traditional presentation, that's why the bufx exist.
I have changed the final compare to ASSERT_COMPARE().

File: tests/suites/test_suite_ecp.data [hunk 9]
```diff
@@ -1027,6 +1027,8 @@ ecp_setup_negative_test #28 Invalid Curve Type
 depends_on:MBEDTLS_ECP_DP_CURVE448_ENABLED
 ecp_mod_setup:"fffffffffffffffffffffffe26f2fc17f69466a74defd8d":MBEDTLS_ECP_DP_NONE:MBEDTLS_ECP_MOD_SCALAR:MBEDTLS_ERR_ECP_BAD_INPUT_DATA
 
+# The following data be generated by random.getrandbits() in python must less than the named curves' modulus.
```
[the remark this answers is not in the document]
Comment (xkqian) [reply tests/suites/test_suite_ecp.data:1030]: That makes sense.
Should we generate another copy of data with seed() be setted in this PR? if so, I will generate another series of data.

[the remark this answers is not in the document]
Comment (yanesca) [reply tests/suites/test_suite_ecp.data:1030]: It is not a blocker, we don't have to, but you can if you feel like it.

File: tests/suites/test_suite_ecp.function [hunk 10]
```diff
@@ -1550,3 +1550,59 @@ exit:
...
+    mbedtls_mpi_mod_residue rA; // For input
+    mbedtls_mpi_mod_residue rA_inverse; // For inverse input
+    mbedtls_mpi_mod_residue rX; // For result
+    mbedtls_mpi_uint *rX_raw = NULL;
+    mbedtls_mpi_uint *A_inverse = NULL;
+    mbedtls_mpi_uint *A = NULL;
+    mbedtls_mpi_uint *bufx = NULL;
+    const mbedtls_mpi_uint one[2] = {1, 0};
+
+    ((void) iret);
```
Comment (yanesca) [inline tests/suites/test_suite_ecp.function:1568]: I think we should remove this from the parameter list entirely, we don't need this.

File: tests/suites/test_suite_ecp.function [hunk 11]
```diff
@@ -1550,3 +1550,59 @@ exit:
...
+    /* Get inverse of A mode m, and multiply it with itself,
+        * to see whether the result equal to '1' */
+    TEST_EQUAL(0, mbedtls_mpi_mod_inv(&rA_inverse, &rA, &m));
+    TEST_EQUAL(mbedtls_mpi_mod_mul(&rX, &rA, &rA_inverse, &m), 0);
+
+    ASSERT_ALLOC(bufx, limbs);
+    TEST_EQUAL(mbedtls_mpi_mod_write(&rX, &m, (unsigned char *) bufx,
+                                        limbs * sizeof(mbedtls_mpi_uint),
+                                        MBEDTLS_MPI_MOD_EXT_REP_LE), 0);
+    ASSERT_COMPARE(bufx, 2, one, 2);
```
Comment (yanesca) [inline tests/suites/test_suite_ecp.function:1595]: This is correct if `rX` has two limbs, but can accept numbers that are not one if they have more than two limbs. What I was trying to suggest that we keep the comparison for the least significant limb as it was and separately we check and make sure that all the remaining limbs are 0.

**Review by yanesca — CHANGES_REQUESTED** [review]
It is looking better now, there has been two occasions where the new version doesn't fix my remarks entirely and I left new comments about these.

**Review by yanesca — APPROVED** [review]
LGTM

**Review by minosgalanakis — APPROVED** [review]
Looks good

**Pull request merged**